### PR TITLE
[android] - Refactor dependencies, introduce focused components

### PR DESCRIPTION
**Pull request by tobrun**
WIP to refactor all gesture related code into a separate wrapper.
- [x] extract code from MapView and make it work
- [x] minimize internal dependencies on GestureDectector
- [x] remove MapView related methods, talk to NativeMapView directly.

**update**: intent has changed from initial creation see below

### CODE REVIEW

**Comment by ivovandongen** [comment]
@tobrun There are indeed a lot of inter-dependecies. And also some 2-way dependencies. It might be helpful for the clarity of the architecture to abstract some things a little bit more. For example, 

- The `MapGestureDetector` has a dependency on MapView, which actually is only passed to the `ZoomButtonsController` as far as I see which only expects a `View`, not a concrete class like `MapView`.  
- In case of the `MapboxMap` dependency, it's actually the `Transform`, `Projection` and `UISettings` you're after.
- The `NativeMapView` dependency is needed for 2 things; to get the pixelRatio/pitch and to control and to effect the gestures on the Map. You might want to abstract the second repsonsibility in an Interface.

There might be a couple of places where interfaces/choosing the highest applicable abstraction might help to clear up the dependencies a bit.

**Comment by tobrun** [comment]
> The MapGestureDetector has a dependency on MapView, which actually is only passed to the ZoomButtonsController as far as I see which only expects a View, not a concrete class like MapView.

Thinking about this again, the zoombutton controller is more like a widget shown on top of the map, I moved this back to MapView with the other widgets. These will be extracted as part of #6570.

>    In case of the MapboxMap dependency, it's actually the Transform, Projection and UISettings you're after.

I tried to limit the amount of arguments but not assigning mapbox to field in the MapGestureDetector. I have now changed this to take each dependency seperatly. 

 
>  The NativeMapView dependency is needed for 2 things; to get the pixelRatio/pitch and to control and to effect the gestures on the Map. You might want to abstract the second repsonsibility in an Interface.

I'm now working to remove the NativeMapView in total, the second responsibility should be made a part of the transform class since they transform the map.

Thanks for some initial :eyes:, will ping again for second round.

**Comment by tobrun** [comment]
The goal of this PR has diverged from it's initial intent, changes made cascaded in changes elsewhere. This PR is now more focused on a general refactor and taking apart the big dependencies into smaller single focused dependencies. This PR adds a lot of todo to the code but current state is acceptable to merge. (we are only missing the snapshot feature, the jni code needs to be refactored and can be picked up with #6238). Focus of Review would be MapboxMap, MapGestureDetector, MapKeyListener, Transform, TrackingSettings and UiSettings. To make the refactor possible I had to remove some unit tests.

File: platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/annotations/MarkerViewManager.java [hunk 1]
```diff
@@ -31,27 +32,33 @@
...
      *
-     * @param mapboxMap the MapboxMap associated with the MarkerViewManager
-     * @param mapView   the MapView associated with the MarkerViewManager
+     * @param container the ViewGroup associated with the MarkerViewManager
      */
-    public MarkerViewManager(@NonNull MapboxMap mapboxMap, @NonNull MapView mapView) {
-        this.mapboxMap = mapboxMap;
+    public MarkerViewManager(@NonNull ViewGroup container) {
+        this.markerViewContainer = container;
         this.markerViewAdapters = new ArrayList<>();
```
Comment (ivovandongen) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/annotations/MarkerViewManager.java]: The initializations that have no dependencies might as well be done in the field declaration. That also enables us to make them final and maybe skip some null checks here and there.

File: platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/AnnotationManager.java [hunk 2]
```diff
@@ -33,29 +36,39 @@
...
     private MapboxMap.OnMarkerClickListener onMarkerClickListener;
     private boolean isWaitingForRenderInvoke;
 
-    AnnotationManager(NativeMapView view, IconManager iconManager, InfoWindowManager manager) {
+    AnnotationManager(NativeMapView view, MapView mapView, MarkerViewManager markerViewManager) {
         this.nativeMapView = view;
-        this.iconManager = iconManager;
-        this.infoWindowManager = manager;
+        this.iconManager = new IconManager(nativeMapView);
+        this.infoWindowManager = new InfoWindowManager();
```
Comment (ivovandongen) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/AnnotationManager.java]: Initialize in field declaration

File: platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/AnnotationManager.java [hunk 3]
```diff
@@ -33,29 +36,39 @@
...
-
+        this.mapView = mapView;
+        this.markerViewManager = markerViewManager;
         if (view != null) {
             // null checking needed for unit tests
             view.addOnMapChangedListener(this);
         }
     }
 
+    // TODO refactor MapboxMap out for Projection and Transform
```
Comment (ivovandongen) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/AnnotationManager.java]: Let's note these things on #6912 as well. I'm planning to pick that up soon.

File: platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/MapGestureDetector.java [hunk 4]
```diff
@@ -0,0 +1,598 @@
...
+     * <p>
+     * Examples of such events are mouse scroll events, mouse moves, joystick & trackpad.
+     * </p>
+     *
+     * @param event The MotionEvent occured
+     * @return True is the event is handled
+     */
+    boolean onGenericMotionEvent(MotionEvent event) {
+        // Mouse events
+        //if (event.isFromSource(InputDevice.SOURCE_CLASS_POINTER)) { // this is not available before API 18
```
Comment (ivovandongen) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/MapGestureDetector.java]: Stray comment?

File: platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/MapView.java [hunk 5]
```diff
@@ -180,13 +137,14 @@ public MapView(@NonNull Context context, @Nullable MapboxMapOptions options) {
...
+    private void initialize(@NonNull final Context context, @NonNull final MapboxMapOptions options) {
         if (isInEditMode()) {
             // if we are in an editor mode we show an image of a map
             LayoutInflater.from(context).inflate(R.layout.mapbox_mapview_preview, this);
             return;
         }
 
+        // TODO distill into singular purpose methods/classes
         initialLoad = true;
         onMapReadyCallbackList = new ArrayList<>();
```
Comment (ivovandongen) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/MapView.java]: Initialize in declaration

File: platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/AnnotationManager.java (hunk 3 above)
Comment (tobrun) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/AnnotationManager.java]: https://github.com/mapbox/mapbox-gl-native/issues/6912#issuecomment-264445033

File: platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/MapView.java (hunk 5 above)
Comment (tobrun) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/MapView.java]: :white_check_mark:

File: platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/AnnotationManager.java (hunk 2 above)
Comment (tobrun) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/AnnotationManager.java]: :white_check_mark:

File: platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/annotations/MarkerViewManager.java (hunk 1 above)
Comment (tobrun) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/annotations/MarkerViewManager.java]: :white_check_mark:

File: platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/MapGestureDetector.java (hunk 4 above)
Comment (tobrun) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/MapGestureDetector.java]: it seems not.. after that line it indicates:  `// this is not available before API 18`, so it's  there waiting for us to bump the minSDK requirments? IDK, this code is from before my time.

**Comment by tobrun** [comment]
I have worked in the requested changes and working now on getting the MapboxMapTest back up and running since this contains our public API

**Comment by tobrun** [comment]
I got the important tests back to run by using mockito 2 configuration for mocking final classes. Though I wasn't able to bring the MapboxMap test since they rely on NativeMapView (which relies on loading the native library). I'm thinking of porting those tests to an instrumentation test equivalent. I think we are covered for now. @ivovandongen can you do another review round?

**Comment by ivovandongen** [comment]
> I wasn't able to bring the MapboxMap test since they rely on NativeMapView (which relies on loading the native library). I'm thinking of porting those tests to an instrumentation test equivalent. I think we are covered for now. 

Let's open a follow-up ticket for that then.

File: platform/android/MapboxGLAndroidSDKTestApp/src/test/java/com/mapbox/mapboxsdk/maps/MapboxMapTest.java [hunk 6]
```diff
@@ -31,512 +34,482 @@
 import static org.mockito.Mockito.when;
 
 public class MapboxMapTest {
-
```
Comment (ivovandongen) [inline platform/android/MapboxGLAndroidSDKTestApp/src/test/java/com/mapbox/mapboxsdk/maps/MapboxMapTest.java]: Why not just use @ignore on the class for now? Or does that make it crash. Makes it more visible that we need to do some work here.

**Comment by tobrun** [comment]
@ivovandongen I have actually now have ported all the unit MapboxMapTest to an instrumentation equivalent. Can you review?

**Comment by tobrun** [comment]
I'm running the Android job multiple times on Bitrise to ensure we aren't introducing any flaky tests.

**Comment by tobrun** [comment]
Not seeing any immediate fixes, going to ignore the MapboxMapTest and create a follow up ticket.

**Pull request merged**